### DISCUSSION TITLE
KQL properties vs child nodes

### DISCUSSION THREAD

**Discussion by nichtich** [kdl-org/kdl#230] [Q&A] [answered]
Given this KDL document

~~~
package name="foo" {
    name "bar"
}
~~~

What does KQL `package name` return? 

* the value `"foo"`
* the value `"bar"`
* the values `"foo"` and `"bar"` (in this order)
* the values `"bar"` and `"foo"` (other way round)

Same question for `package [] name` (select `package` element, then "any element", then value of "name").

The result should be the same for `package => prop(name)` and `package => name`, right?

For `package > bar` the result should be the value `"bar"`.

At least for `package + name` and `package ~ name` the result would be empty.

**Comment by zkat** [comment, marked as answer]
(note, we're working on new syntax, and `>>` is how we refer to descendants now. `++` is the same but for siblings)

Ok so.

KQL, with regular (non-map) syntax, only returns _nodes_. That is, `package >> name` and `package > name` will return the `name "bar"` node itself. Ditto for ANY selectors that select `name`.

The map operator, by the way, is going away with 2.0, but if you're implementing the 1.0 operator and want something like it, then yes, `=> prop(name)` and `=> name` are the same thing.

This is part of the reason I want to remove the map operator in 2.0: I really want people to be clear that what you're selecting is always _nodes_, not their properties or values. Think `document.querySelector()`.

**Reply by nichtich** [reply]
Thanks and good to hear there will be a revised version of KQL. My next question would have been the kind of elements that KQL can return:

> I really want people to be clear that what you're selecting is always *nodes*, not their properties or values. Think `document.querySelector()`.

Or `document.querySelectorAll()`. With KQL you always select another KDL document (= a sequence of nodes) from a KDL document, don't you? Nice!
